chainindex/api.go

```diff
@@ -126,17 +127,71 @@ func (si *SqliteIndexer) ChainValidateIndex(ctx context.Context, epoch abi.Chain
 		return nil, xerrors.Errorf("failed to cast tipset key cid: %w", err)
 	}
 
-	if indexedTsKeyCid != expectedTsKeyCid {
+	if !indexedTsKeyCid.Equals(expectedTsKeyCid) {
 		return nil, xerrors.Errorf("index corruption: non-reverted tipset at height %d has key %s, but canonical chain has %s", epoch, indexedTsKeyCid, expectedTsKeyCid)
 	}
 
+	indexedNonRevertedMsgCount, indexedNonRevertedEventsCount, hasRevertedEvents, err := si.getIndexedTipSetData(ctx, expectedTs.Key())
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset message and event counts at height %d: %w", expectedTs.Height(), err)
+	}
+
+	if err = si.verifyIndexedData(ctx, expectedTs, indexedNonRevertedMsgCount); err != nil {
+		return nil, xerrors.Errorf("failed to verify indexed data at height %d: %w", expectedTs.Height(), err)
+	}
+
 	return &types.IndexValidation{
-		TipsetKey: expectedTs.Key().String(),
-		Height:    uint64(expectedTs.Height()),
-		// TODO Other fields
+		TipsetKey:      expectedTs.Key().String(),
+		Height:         uint64(expectedTs.Height()),
+		TotalMessages:  indexedNonRevertedMsgCount,
+		TotalEvents:    indexedNonRevertedEventsCount,
+		EventsReverted: hasRevertedEvents,
 	}, nil
 }
 
+// verifyIndexedData verifies that the indexed data for a tipset is correct
+// by comparing the number of messages in the chainstore to the number of messages indexed
+
+// TODO: verify indexed events too (to verify the events we need to load the next tipset (ts+1) and verify the events are the same)
```

Comment:
Please can you implement this TODO to verify events as well ?

Reply:
You can use the `loadExecutedMessages` function we have for this. Then you can just read messages and events together.

Reply:
I was little skeptical about events, because we need to do something like `currTipset.Height() + 1` to get execution tipset then fetch data for it through `loadExecuteMessages`.

So should I go ahead with this?

Reply:
@akaladarshi 

Hmm..good point.

Yeah let's do it. So, here are the steps:

1) Load tipset `tsX` at user supplied epoch `e`
2) Load tipset `tsY` at epoch `e+1`
3) Ensure parent of ``tsY` == `tsX` (if there is a mismatch here, return an error because it means chain has forked and user needs to retry)
4) Load events and messages and validate the counts.